### ISSUE TITLE
Current issues with `Value`

### ISSUE THREAD

**Issue by torkleyy** [ron-rs/ron#122]
* [x] tuples / tuple structs not supported (#121)
* [ ] enums not supported
* [ ] struct names don't get stored
* [ ] units represented with struct names
* [ ] Make sure errors make sense, esp after a unit struct (e.g. `UnitStruct:` as input)
* [ ] figure out what to do about tuples becoming sequences

**Comment by Boscop** [comment]
Can some of these boxes be ticket now?

**Comment by kvark** [comment]
comment

**Comment by wiiznokes** [comment]
Any plans on improving this or this is out of scope ?

**Comment by juntyr** [comment]
Yes, many plans, little time (research + small child) :/ If you can help with any part, I'd really appreciate it :)

**Comment by wiiznokes** [comment]
> Yes, many plans, little time (research + small child) :/ If you can help with any part, I'd really appreciate it :)

I'm interested in making the conversion "ron str" -> `ron::Value` lossless. Maybe i could try to implement it, but i think it would help if you can give some hint about the change needed in the Value type.

**Comment by juntyr** [comment]
> > Yes, many plans, little time (research + small child) :/ If you can help with any part, I'd really appreciate it :)
> 
> I'm interested in making the conversion "ron str" -> `ron::Value` lossless. Maybe i could try to implement it, but i think it would help if you can give some hint about the change needed in the Value type.

Essentially the Value type would need to become an AST-like type for everything the current grammar can represent. Then ron string -> Value -> string would be lossless.

You can find the grammar here: https://github.com/ron-rs/ron/blob/master/docs/grammar.md (it even has a Value section).

**Comment by wiiznokes** [comment]
Ok, i've got smtg like this:

```rs
pub enum Value {
    Bool(bool),
    Char(char),
    Map(Map),
    Struct(Option<String>, Map),
    Number(Number),
    Option(Option<Box<Value>>),
    String(String),
    Bytes(Vec<u8>),
    Seq(Vec<Value>),
    Tuple(Option<String>, Vec<Value>),
    Unit,
}
```

**Comment by wiiznokes** [comment]
> we also need a named unit variant with a string field (e.g. for unit structs and unit variants)

I was thinking of using `Struct(None, empty::map)` for "()", and `Struct(Some("name"), empty::map)` for unit_struct with indent.
For enum_variant_unit, we can use `Tuple(Some("name"), empty::vec)`.

Do you still think it's better to add a new variant ? In that case, maybe `UnitStruct(Option<String>)`

**Comment by HeitorAugustoLN** [comment]
I want to make my project [cosmic-ctl](https://github.com/cosmic-utils/cosmic-ctl), work with RON values directly instead of a string, but this is currently not possible due to the lack of enum support in `Value` and tuples turning into Vec. I would really appreciate it if this were fixed
